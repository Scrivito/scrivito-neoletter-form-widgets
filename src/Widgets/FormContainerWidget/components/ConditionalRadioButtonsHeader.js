import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../utils/getFieldName";
import { SelectItem } from "./SelectComponent";
import { Mandatory } from "./MandatoryComponent";
import { HelpText } from "./HelpTextComponent";

export const RadioButtonsHeader = Scrivito.connect(
  ({ widget, onChangeSelected }) => {
    return (
      <div
        className={`mb-3 select-container ${
          Scrivito.isInPlaceEditingActive() ? "condition-wrapper" : ""
        }`}
      >
        <div className="select-title">
          <span className="text-super">{widget.get("title")}</span>
          {widget.get("required") && <Mandatory />}
          {widget.get("helpText") && <HelpText widget={widget} />}
        </div>
        <div className="row ">
          {widget.get("conditions").map((condition, index) => (
            <SelectItem
              selectionType={"radio"}
              name={getFieldName(widget)}
              onChange={onChangeSelected}
              value={condition.get("title")}
              id={condition.id()}
              required={widget.get("required")}
              key={index}
            />
          ))}
        </div>
      </div>
    );
  }
);
