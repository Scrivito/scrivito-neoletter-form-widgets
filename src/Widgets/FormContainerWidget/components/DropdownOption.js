import * as React from "react";

export const DropdownOption = ({ value, id }) => {
  return (
    <option value={value} id={id}>
      {value}
    </option>
  );
};
