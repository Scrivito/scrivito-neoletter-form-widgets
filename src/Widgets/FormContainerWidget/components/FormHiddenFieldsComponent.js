import * as React from "react";
import * as Scrivito from "scrivito";
import { getHistory } from "../../../config/history";

export const FormHiddenFields = ({ widget }) => {
  const [browserLocation, setBrowserLocation] = React.useState(null);
  React.useEffect(() => {
    const history = getHistory();
    console.log("And the history is....");
    console.log(history);
    if (!history) return;
    setBrowserLocation(locationToUrl(history.location));

    return history.listen(({ location }) =>
      setBrowserLocation(locationToUrl(location))
    );
  }, []);
  return (
    <>
      <input type="hidden" name="form_id" value={widget.get("formId")} />
      <input
        type="hidden"
        name="url"
        value={browserLocation || Scrivito.urlFor(widget.obj())}
      />
      <Scrivito.InPlaceEditingOff>
        <Scrivito.ContentTag content={widget} attribute="hiddenFields" />
      </Scrivito.InPlaceEditingOff>

      <HoneypotField />
    </>
  );
};

const HoneypotField = () => (
  <div aria-hidden="true" className="winnie-the-pooh">
    <input autoComplete="off" name="fax" tabIndex="-1" />
  </div>
);

function locationToUrl(location) {
  return `${window.location.origin}${location.pathname}${location.search}`;
}
