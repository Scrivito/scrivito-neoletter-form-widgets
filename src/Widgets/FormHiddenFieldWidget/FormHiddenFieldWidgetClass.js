import * as Scrivito from "scrivito";

export const FormHiddenFieldWidget = Scrivito.provideWidgetClass(
  "FormHiddenFieldWidget",
  {
    onlyInside: ["FormContainerWidget"],
    attributes: {
      customFieldName: "string",
      hiddenValue: "string",
    },
  }
);
