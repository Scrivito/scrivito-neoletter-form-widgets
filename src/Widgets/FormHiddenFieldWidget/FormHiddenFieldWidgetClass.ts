import * as Scrivito from "scrivito";

export const FormHiddenFieldWidget = Scrivito.provideWidgetClass(
  "FormHiddenFieldWidget",
  {
    onlyInside: [
      "FormStepContainerWidget",
      "FormStepWidget",
      "FormContainerWidget"
    ],
    attributes: {
      customFieldName: "string",
      hiddenValue: "string",
      type: [
        "enum",
        {
          values: ["custom", "subscription"]
        }
      ]
    }
  }
);
