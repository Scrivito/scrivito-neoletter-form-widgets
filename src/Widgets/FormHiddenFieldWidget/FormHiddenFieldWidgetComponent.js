import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";

Scrivito.provideComponent("FormHiddenFieldWidget", ({ widget }) => {
  const name = getFieldName(widget);
  if (!name) return null;

  return <input type="hidden" name={name} value={widget.get("hiddenValue")} />;
});
