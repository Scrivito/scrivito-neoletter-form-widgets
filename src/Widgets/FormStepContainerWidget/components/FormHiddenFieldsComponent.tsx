import * as React from "react";
import * as Scrivito from "scrivito";
import { Location } from "history";
import { getHistory } from "../../../config/history";

interface FormHiddenFieldsProps {
  widget: Scrivito.Widget;
}
export const FormHiddenFields: React.FC<FormHiddenFieldsProps> = ({
  widget
}) => {
  const [browserLocation, setBrowserLocation] = React.useState<string | null>(
    null
  );
  React.useEffect(() => {
    const history = getHistory();
    if (!history) return;
    setBrowserLocation(locationToUrl(history.location));

    return history.listen(({ location }) =>
      setBrowserLocation(locationToUrl(location))
    );
  }, []);
  return (
    <>
      <input
        type="hidden"
        name="form_id"
        value={widget.get("formId") as string}
      />
      <input
        type="hidden"
        name="url"
        value={browserLocation || Scrivito.urlFor(widget.obj())}
      />
      <Scrivito.InPlaceEditingOff>
        <Scrivito.ContentTag content={widget} attribute="hiddenFields" />
      </Scrivito.InPlaceEditingOff>

      <HoneypotField />
    </>
  );
};

const HoneypotField = () => (
  <div aria-hidden="true" className="winnie-the-pooh">
    <input autoComplete="off" role="presentation" name="fax" tabIndex={-1} />
  </div>
);

function locationToUrl(location: Location) {
  return `${window.location.origin}${location.pathname}${location.search}`;
}
