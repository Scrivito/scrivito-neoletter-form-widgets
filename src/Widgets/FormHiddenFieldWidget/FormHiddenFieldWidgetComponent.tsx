import * as React from "react";
import * as Scrivito from "scrivito";
import { isEmpty } from "lodash-es";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { FormHiddenFieldWidget } from "./FormHiddenFieldWidgetClass";

Scrivito.provideComponent(FormHiddenFieldWidget, ({ widget }) => {
  const name = getFieldName(widget);
  if (isEmpty(name)) return null;
  return <input type="hidden" name={name} value={widget.get("hiddenValue")} />;
});
