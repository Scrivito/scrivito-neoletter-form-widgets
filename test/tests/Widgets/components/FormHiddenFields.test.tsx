import * as Scrivito from "scrivito";
import * as React from "react";
import { render } from "@testing-library/react";
import { FormHiddenFields } from "../../../../src/Widgets/FormStepContainerWidget/components/FormHiddenFieldsComponent";
import { DummyWidget } from "../../../helpers/dummyWidget";
import { FormHiddenFieldWidget } from "../../../../src/Widgets/FormHiddenFieldWidget/FormHiddenFieldWidgetClass";
import "../../../../src/Widgets/FormHiddenFieldWidget/FormHiddenFieldWidgetComponent";
import PageRenderer from "../../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

describe("FormHiddenFields", () => {
  const formId = "myFormId";
  const widget = new DummyWidget({
    formId
  }) as unknown as Scrivito.Widget;

  it("renders the form_id input field", () => {
    const { container } = render(<FormHiddenFields widget={widget} />);
    const formIdInput = container.querySelector('input[name="form_id"]');

    expect(formIdInput).toBeInTheDocument();
    expect(formIdInput).toHaveAttribute("type", "hidden");
    expect(formIdInput).toHaveAttribute("value", formId);
  });

  it("renders the url input field", () => {
    const { container } = render(<FormHiddenFields widget={widget} />);
    const urlInput = container.querySelector('input[name="url"]');

    expect(urlInput).toBeInTheDocument();
    expect(urlInput).toHaveAttribute("type", "hidden");
    expect(urlInput).toHaveAttribute("value", "http://localhost/");
  });

  it("renders the HoneypotField", () => {
    const { container } = render(<FormHiddenFields widget={widget} />);
    const honeypotField = container.querySelector(
      '.winnie-the-pooh input[name="fax"]'
    );
    expect(honeypotField).toHaveAttribute("tabIndex", "-1");
    expect(honeypotField).toHaveAttribute("role", "presentation");
  });

  it("renders the hiddenFields", () => {
    const pageRenderer = new PageRenderer();
    pageRenderer.render({
      body: [
        new FormHiddenFieldWidget({
          customFieldName: "custom_hidden1",
          hiddenValue: "a hidden val",
          type: "custom"
        }),
        new FormHiddenFieldWidget({
          customFieldName: "custom_hidden2",
          hiddenValue: "another hidden val",
          type: "custom"
        })
      ]
    });
    const hiddenFields = document.querySelectorAll("input");
    expect(hiddenFields).toHaveLength(2);
    const expectedValues = ["a hidden val", "another hidden val"];
    hiddenFields.forEach((input, index) => {
      expect(input).toHaveAttribute("type", "hidden");
      expect(input).toHaveAttribute("value", expectedValues[index]);
    });
  });

  it("renders correctly", () => {
    const pageRenderer = new PageRenderer();
    const tree = pageRenderer.getAsJSON({
      body: [
        new FormHiddenFieldWidget({
          customFieldName: "custom_hidden1",
          hiddenValue: "a hidden val"
        }),
        new FormHiddenFieldWidget({
          customFieldName: "custom_hidden2",
          hiddenValue: "another hidden val"
        })
      ]
    });
    expect(tree).toMatchSnapshot();
  });
});
