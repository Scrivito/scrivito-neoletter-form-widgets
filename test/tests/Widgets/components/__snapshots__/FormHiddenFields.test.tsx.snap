// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHiddenFields renders correctly 1`] = `
<div>
  <div>
    <input
      name={null}
      type="hidden"
      value="a hidden val"
    />
  </div>
  <div>
    <input
      name={null}
      type="hidden"
      value="another hidden val"
    />
  </div>
</div>
`;
